### PR TITLE
Alyapunov/gh 6132 mvcc lost tuple after conflict exception v4

### LINKED ISSUE #6021: Incorrect rollback with delete operation in MVCC
**Bug description**

Here is a test of that bug:

```lua
box.cfg { memtx_use_mvcc_engine = true }

s = box.schema.create_space('test')
s:create_index('pk', {parts={{1, 'uint'}}})
s:create_index('sec', {parts={{2, 'uint'}}})

fiber = require("fiber")
s:insert({1, 0})

box.begin()
s:delete(1)

fiber.create(function()
    box.begin()
    s:replace({1,1})
    fiber.sleep(1)
    box.commit()
end)

fiber.sleep(1)
box.commit()
```

After that, I get strange errors and tarantool exits with the next messages:

> 2021-04-19 16:04:45.890 [80786] main/103/interactive memtx_tree.cc:863 E> ER_TUPLE_FOUND: Duplicate key exists in unique index "pk" in space "test" with old tuple - [1, 1] and new tuple - [1, 0]
[1]    80786 illegal hardware instruction  src/tarantool

* OS: MacOS
* OS Version: MacOS 11.0.1
* Architecture: amd64

Tarantool 2.5.2-0-g05730d3
Target: Darwin-x86_64-Release
Build options: cmake . -DCMAKE_INSTALL_PREFIX=/usr/local/Cellar/tarantool/2.5.2.0 -DENABLE_BACKTRACE=ON
Compiler: /usr/bin/clang /usr/bin/clang++
C_FLAGS: -fexceptions -funwind-tables -fno-omit-frame-pointer -fno-stack-protector -fno-common -fopenmp -msse2 -std=c11 -Wall -Wextra -Wno-strict-aliasing -Wno-char-subscripts -Wno-format-truncation -Wno-gnu-alignof-expression -Wno-cast-function-type
CXX_FLAGS: -fexceptions -funwind-tables -fno-omit-frame-pointer -fno-stack-protector -fno-common -fopenmp -msse2 -std=c++11 -Wall -Wextra -Wno-strict-aliasing -Wno-char-subscripts -Wno-format-truncation -Wno-invalid-offsetof -Wno-gnu-alignof-expression -Wno-cast-function-type

### LINKED ISSUE #6132: MVCC: loss of tuples after a conflict exception
* OS: Linux, Debian
* OS Version: bullseye/sid
* Architecture: amd64
* Version: 2.9.0-61-g2b25ca6c7

Condition to reproduce: use MVCC in background in the other transactions.

```text
$ rm -f reproduce/00000*; tarantool reproduce.lua
...
record is in database
record is in database
transaction finish status:      false   Transaction has been aborted by conflict
record is NOT in database
```

Reproduce script:

```lua
console = require 'console'
fiber = require 'fiber'
netbox = require 'net.box'

math.randomseed(os.time())

box.cfg{
    work_dir = 'reproduce',
    memtx_use_mvcc_engine = true,
    listen = 3301,
}


box.once(
    'init-schema',
    function()
        box.schema.user.create('test', {password = 'test'})
        box.schema.user.grant('test', 'execute,read,write', 'universe')


        box.schema.space.create(
            'accounts',
            {
                format = {
                    {   name    = 'user', type  = 'string' },
                    {   name    = 'balance', type = 'integer' },
                }
            }
        )

        box.space.accounts:create_index(
            'pk',
            {
                parts = {
                    { field = 1, type='string' },
                }
            }
        )

        box.space.accounts:insert{ 'vasya', 0 }
        box.space.accounts:insert{ 'petya', 0 }
    end
)

accounts = box.space.accounts

local function background_mvcc()
    while true do
        box.begin()
        local p = accounts:get('petya')
        local add = math.ceil(math.random() * 200) - 100
        accounts:update('petya', {{'=', 'balance', p['balance'] + add}})
        box.commit()
    end
end

local function record_status()
    if accounts:get 'vasya' then
        print 'record is in database'
    else
        print 'record is NOT in database'
    end
end


record_status()
fiber.create(background_mvcc)   -- Condition to reproduce!

nbx = netbox.connect('test:test@localhost:3301')

record_status()

box.begin()
accounts:update('vasya', {{'=', 'balance', 10}})
nbx.space.accounts:update('vasya', {{'=', 'balance', -10}})
fiber.sleep(0.1)

print('transaction finish status:', pcall(box.commit, {}))

record_status()

console.start()
os.exit()

```

### PR DESCRIPTION
**Pull request by alyapunov**
txm: rewrite and refactor mvcc code
    
There were several problems that was connected with broken
pointers in tuple history. Another problems is that that code
was quite huge and difficult to understand.
    
This patch refactors all the code that is connected to lists of
stories in history. A bunch of helper function was added and in
fact these functions was carefully rewtitten:
 * memtx_tx_history_add_stmt
 * memtx_tx_history_rollback_stmt
 * memtx_tx_history_prepare_stmt
 * memtx_tx_history_commit_stmt
    
In addition to refactoring a couple of significant changes was
made to the logic:
 * Now del_story in statement point to story of the tuple that
was effectively deleted by this statement.
 * Conflicts in secondary indexes (that were previously named as
'cross coflicts' now handled transparently during statement
preparation.
    
Closes #6132
Closes #6021

### CODE REVIEW

File: src/box/memtx_tx.c [hunk 1]
```diff
@@ -1129,84 +1305,54 @@ memtx_tx_history_add_stmt(struct txn_stmt *stmt, struct tuple *old_tuple,
...
-		stmt->next_in_del_list = del_story->del_stmt;
-		del_story->del_stmt = stmt;
-		stmt->del_story = del_story;
+	directly_replaced_count = space->index_count;
+	struct tuple *replaced = directly_replaced[0];
+
+	/* Check overwritten tuple */
+	int rc;
+	if (replaced == NULL || !replaced->is_dirty)
+		rc = check_dup_clean(stmt, new_tuple, directly_replaced,
```
Comment (Korablev77) [inline src/box/memtx_tx.c:1327]: check_dup_clean and check_dup dirty have the same signature. Mb it is worth placing dispatching inside single check_dup()?

Comment (alyapunov) [reply src/box/memtx_tx.c:1327]: The idea was to split code into parts.
Or you mean make one `check_dup` that would call either `check_dup_clean` or `check_dup_dirty`?

Comment (Korablev77) [reply src/box/memtx_tx.c:1327]: Yep

Comment (alyapunov) [reply src/box/memtx_tx.c]: done

**Review by EvgenyMekhanik — APPROVED** [review]

**Review by Korablev77 — APPROVED** [review]
LGTM

**Comment by Korablev77** [comment]
Already pushed to master.

**Pull request closed without merging**